apps/kg/eval.py

```diff
@@ -105,102 +94,74 @@ def main(args):
-                                                        num_workers=num_workers,
-                                                        rank=0, ranks=1)
-        test_sampler_tail = eval_dataset.create_sampler('test', args.batch_size,
-                                                        args.neg_sample_size,
-                                                        args.neg_chunk_size,
+                                                        num_workers=1,
+                                                        rank=i, ranks=args.num_proc)
+        test_sampler_tail = eval_dataset.create_sampler('test', args.batch_size_eval,
+                                                        args.neg_sample_size_test,
+                                                        args.neg_sample_size_test,
                                                         args.eval_filter,
                                                         mode='chunk-tail',
-                                                        num_workers=num_workers,
-                                                        rank=0, ranks=1)
+                                                        num_workers=1,
+                                                        rank=i, ranks=args.num_proc)
+        test_sampler_heads.append(test_sampler_head)
+        test_sampler_tails.append(test_sampler_tail)
 
     # load model
     n_entities = dataset.n_entities
     n_relations = dataset.n_relations
     ckpt_path = args.model_path
     model = load_model_from_checkpoint(logger, args, n_entities, n_relations, ckpt_path)
 
-    if args.num_proc > 1:
-        model.share_memory()
+    model.share_memory()
     # test
     args.step = 0
     args.max_step = 0
     start = time.time()
-    if args.num_proc > 1:
-        queue = mp.Queue(args.num_proc)
-        procs = []
-        for i in range(args.num_proc):
-            proc = mp.Process(target=test_mp, args=(args,
-                                                    model,
-                                                    [test_sampler_heads[i], test_sampler_tails[i]],
-                                                    i,
-                                                    'Test',
-                                                    queue))
-            procs.append(proc)
-            proc.start()
-
-        total_metrics = {}
-        metrics = {}
-        logs = []
-        for i in range(args.num_proc):
-            log = queue.get()
-            logs = logs + log
-
-        for metric in logs[0].keys():
-            metrics[metric] = sum([log[metric] for log in logs]) / len(logs)
-        for k, v in metrics.items():
-            print('Test average {} at [{}/{}]: {}'.format(k, args.step, args.max_step, v))
-
-        for proc in procs:
-            proc.join()
-    else:
-        test(args, model, [test_sampler_head, test_sampler_tail])
+    #if args.num_proc > 1:
```

Comment:
let's not always create a new process to run eval. It's more difficult to debug when running the computation in a different process.

Reply:
Agree, if num_proc == 1, we can avoid fork.